flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/copy/HiveParserBaseSemanticAnalyzer.java

```diff
@@ -387,7 +387,9 @@ private static void processPrimaryKeyInfos(HiveParserASTNode pkNode, List<PKInfo
     }
 
     private static void checkColumnName(String columnName) throws SemanticException {
-        if (VirtualColumn.VIRTUAL_COLUMN_NAMES.contains(columnName.toUpperCase())) {
```

Comment:
It's really wired to me. It'wll throw `java.lang.NoSuchFieldError: VIRTUAL_COLUMN_NAMES` with this line [fail](https://github.com/apache/flink/pull/19926) . 
But it does exist in `VirtualColumn`.
And It's fine in Flink 1.15.0, so I believe some changes to master boils it.
So i change this line, and then the change is fine.Sucessful test is in [here](https://github.com/apache/flink/pull/19929)
Also, I validate it by flink sql client.

Reply:
Maybe this is caused by using a different hive version in the compile/test stage?

Reply:
Thanks for your reminder.  But I also compile `flink-sql-connector-hive-connector-3.1.2`, and the same exception throw.
Then, I decompile and check the class file `org.apache.hadoop.hive.ql.metadata.VirtualColumn` in the `flink-sql-connector-hive-connector-3.1.2.jar`. I found the the line `public static final org.apache.flink.hive.shaded.com.google.common.collect.ImmutableSet<java.lang.String>.  However it's different from the type `VirtualColumn.VIRTUAL_COLUMN_NAMES` in this method `checkColumnName`.  So, the excpetion throw.
When replace with it with a method instead of field reference, we can solve such problem.

Reply:
That's because we relocate guava in flink-sql-connector-hive, but that is also applied to the the Flink code bundled in flink-sql-connector-hive, so it shouldn't matter.

We are also still lacking an explanation for why this only fails with Hive 3, and not with Hive 2 (which uses the same code with the same relocation).

Reply:
Sorry for lacking of detail explanation. Yes, the core reson isn't relocated, but is type inconsistent . The difference  in `org.apache.hadoop.hive.ql.metadata.VirtualColumn` between hive2-exec and hive3-exec bring the case.

I decompile `sql-connector-hive-connector-2.3.8.jar`, and the bytecode of the method `checkColumnName ` looks like:
`
Code:
       0: getstatic     #419                // Field org/apache/hadoop/hive/ql/metadata/VirtualColumn.VIRTUAL_COLUMN_NAMES:Lorg/apache/hive/com/google/common/collect/ImmutableSet
`
the type of `VirtualColumn.VIRTUAL_COLUMN_NAMES` should be `org/apache/hive/com/google/common/collect/ImmutableSet`.

And the bytecode of the `VirtualColumn` is also
`
 public static final org.apache.flink.hive.shaded.com.google.common.collect.ImmutableSet<java.lang.String> VIRTUAL_COLUMN_NAMES;
`
So, it's fine with hive2.

Then, I decompile `sql-connector-hive-connector-3.1.2.jar`, and the bytecode of the method `checkColumnName ` looks like:
`
Code:
       0: getstatic     #419                // Field org/apache/hadoop/hive/ql/metadata/VirtualColumn.VIRTUAL_COLUMN_NAMES:Lorg/apache/hive/com/google/common/collect/ImmutableSet
`
the type of `VirtualColumn.VIRTUAL_COLUMN_NAMES` should be `org/apache/hive/com/google/common/collect/ImmutableSet`.

But the bytecode of the `VirtualColumn` is 
`
 public static final org.apache.flink.hive.shaded.com.google.common.collect.ImmutableSet<java.lang.String> VIRTUAL_COLUMN_NAMES;
`
It's relocated by flink, which isn't consistent to `org/apache/hive/com/google/common/collect/ImmutableSet`.
So, `NoSuchFieldError` throws.

Also, I decompile the hive-exec-2.3.9 and hive-exec-3.1.2 that our hive connector depend on. I find in hive-exec-2.3.9, the `VIRTUAL_COLUMN_NAMES` is type of `org.apache.hive.com.google.common.collect.ImmutableSet<java.lang.String>`. But in hive-exec-3.1.2, the the `VIRTUAL_COLUMN_NAMES` is type of `com.google.common.collect.ImmutableSet<java.lang.String>` which is to be relocated to `org.apache.flink.hive.shaded.com.google.common.collect.ImmutableSet<java.lang.String>`.

Reply:
OK that makes sense, but ultimately this should be caused by compiling against hive2 but running against hive3.

In any case, the switch to the method should resolve the issue.

Reply:
Yes, only when I use `javap -private -c -l xx` to decompile it, did I find the difference.
The decompiler provided by IDEA is user friendly,  but miss some important things.

Reply:
Only certain features. Only when use hive dialect to create table, it will throw exception. Otherwise, it works normally.